### PR TITLE
Don't define JENV_DIR from the 1st 'java' file param

### LINKED ISSUE #151: Jenv sometimes decides Java version according to command arguments
I just spent at least an hour fighting with a Jenv induced issue, and I'm trying to understand it. I have:

 * jenv local 1.8
 * jenv global 1.8
 * export JAVA_HOME=<java 1.8 home>
 * java -version => prints java 1.8 version message

And yet I run my program using

```
java -cp /Users/simon/.sdkman/candidates/groovy/2.4.6/embeddable/groovy-all-2.4.6.jar my.Class /Users/foo/bar/test.txt
```

And it runs using *java 1.7*.

After an hour of debugging I figure this out: jenv has the following logic in the shim:

```bash
if [ "$program" = "java" ]; then
  for arg; do
    case "$arg" in
    -e* | -- ) break ;;
    */* )
      if [ -f "$arg" ]; then
        export JENV_DIR="${arg%/*}"
        break
      fi
      ;;
    esac
  done
fi
```

So decoding this, if you are running `java`, jenv looks at all the arguments and if it finds any argument containing a slash then it sets JENV_DIR to the directory of that file. Then the java version is decided by the java version set for that directory rather than the directory you are currently in.

Can someone explain why it's doing this?

### LINKED ISSUE #162: jenv picks up .java-version from argument instead of current directory
### steps

1. `cd ~/foo/`
2. `jenv local 1.8`
3. Suppose there's a jar file under `~/foo/` that I built called `~/foo/target/foo.jar`.
4. `cd ~/java9-test/`
5. `jenv local 9-ea`
6. `java -version` (This should display java version "9-ea" etc..)
7. `java -jar ~/foo/foo.jar`

### problem

Even though there is `.java-version` file directly under `~/java9-test/`, `java -jar ~/foo/foo.jar` runs as Java 8, picking up the file from `~/foo/`.

### expectation

My expectation was that the current directory's `.java-version` is going to be picked up, running Java 9-ea.

### PR DESCRIPTION
**Pull request by dwijnand**
This logic originates from one of the first ever commits of jenv:
c6819eaf6a60dc4e8ba1cb2de48dfb9eba4547b0.

I'm not sure what justifies this behaviour, but there have been at least
two issues reporting this as breaking expectations: #151 and #162.

So I'm proposing removing it with this patch.

Fixes #151. Fixes #162.

### CODE REVIEW

**Comment by claui** [comment]
**tl;dr** I’d recommend a different fix for better compatibility.

> I'm not sure what justifies this behaviour

@dwijnand Allow me to give a little context, taken from @ssadedin’s excellent analysis: the custom `JENV_DIR` logic runs inside the `java` shim. It attempts to detect whether `java` is being run on a main class from a different directory. [It then `cd`s to that directory](https://github.com/jenv/jenv/blob/af89d7800f93fd9d75935a39e5cabf6417e0daee/libexec/jenv#L56) in order to pick up whatever Java version may be `jenv local`ly configured there.

### Why the logic exists

I’m not the original author but from my point of view, the rationale behind the `JENV_DIR` logic may be to manage user expectations in the following scenario:

1. Some Java project has a `.java-version` override created by `jenv local`. Example: `/my/projects/foo/.java-version`

2. User has a custom `/usr/local/bin/foo` Bash shim or alias, which runs `java /my/projects/foo/classes/Main.class`, with `java` resolving to the jEnv shim.

3. When running `foo`, the user expects jEnv to pick up the correct Java version for `/my/projects/foo`.

I feel that expectation is perfectly reasonable. There may be many jEnv installations where users depend on it.

### Why it breaks

On running the `java` shim, if a valid file path with a `/` is passed:

a) as an argument to certain Java options, such as `-cp` or `-classpath`,

or b) as an argument to the main class,

then the logic may mistake that path for the main class (or jar, source, or module). It may then fail to pick up the `.java_version` from the current directory hierarchy. It may also pick up a wrong `.java_version` from an unrelated directory hierarchy. See @ssadedin’s analysis in #151 for more details.

### Proposed fix

We could change the logic to skip all options (`-*`) and their single arguments (e. g., after `-cp`, `-classpath`, `-d`, `-p`, …). The loop should break immediately after consuming the (single) argument to `-jar`, `-m`, and `--module`. It should also break on reaching the first non-option argument (which, according to `java -help`, must be a class or source file). We can then lose both the `[ -f "$arg" ]` and `--` checks.

**Comment by dwijnand** [comment]
Sorry I've been dragging my feet on responding to you.  Your argumentation (that you redacted) made logical sense to me.  It's a design decision on what "local Java version" should mean and what it should impact.  Neither arguments are wrong, per se, in my opinion.

**Comment by claui** [comment]
@dwijnand It just wouldn’t play nice with `java` CLI semantics.

It’s cool as long as we use the `-jar` syntax:

```sh
java -jar /my/projects/foo/lib/test.jar  # this is a case where jenv picks a JENV_DIR
```

Java 12 introduces another convenient mode where you can pass source files directly:

```sh
java /my/projects/foo/src/Test.java  # works on Java 12 and up
```

It’s not so easy though if we run class files. The following is not valid because `java` does not accept class file names:

```sh
java /my/projects/foo/classes/Test.class  # ClassNotFoundException
```

The following is allowed (but `classes` is a directory, which causes the current `JENV_DIR` code to ignore this case):

```sh
java -cp /my/projects/foo/classes Test  # jenv will *not* pick a JENV_DIR here
```

The following is also valid:

```sh
java -cp /my/projects/foo/bar.jar Test  # jenv picks a JENV_DIR here
```

But how to handle this?

```sh
java -cp foo/bar:foo/baz Test  # which JENV_DIR would we even pick?
```

Or this?

```sh
java -cp /usr/local/lib/baz.jar:/my/projects/foo/classes Test  # same here
```

That’s why I’d say it’s more than just a design decision.

**Comment by dwijnand** [comment]
Review by @doctorpangloss or @gcuisinier?

**Comment by dwijnand** [comment]
Happy to resubmit when there's more maintenance bandwidth.
